extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java

```diff
@@ -0,0 +1,296 @@
+    final ServiceLocator controllerLocator = makeControllerLocator(worker.task().getControllerTaskId());
+    periodicControllerCheckerExec.submit(() -> controllerCheckerRunnable(controllerLocator, worker));
+  }
+
+  @VisibleForTesting
+  void controllerCheckerRunnable(final ServiceLocator controllerLocator, final Worker worker)
+  {
+    while (true) {
+      // Add some randomness to the frequency of the loop to avoid requests from simultaneously spun up tasks bunching
+      // up and stagger them randomly
+      long sleepTimeMillis = FREQUENCY_CHECK_MILLIS + ThreadLocalRandom.current().nextLong(
+          -FREQUENCY_CHECK_JITTER,
+          2 * FREQUENCY_CHECK_JITTER
+      );
+      final ServiceLocations controllerLocations;
+      try {
+        controllerLocations = controllerLocator.locate().get();
+      }
+      catch (Throwable e) {
+        // Service locator exceptions are not recoverable.
+        log.noStackTrace().warn(
+            e,
+            "Periodic fetch of controller location encountered an exception. Worker task [%s] will exit.",
+            worker.id()
+        );
+        worker.controllerFailed();
+        break;
+      }
+
+      if (controllerLocations.isClosed() || controllerLocations.getLocations().isEmpty()) {
+        log.warn(
+            "Periodic fetch of controller location returned [%s]. Worker task [%s] will exit.",
+            controllerLocations,
+            worker.id()
+        );
+        worker.controllerFailed();
+        break;
+      }
+
+      try {
+        Thread.sleep(sleepTimeMillis);
+      }
+      catch (InterruptedException ignored) {
+        // Do nothing: an interrupt means we were shut down. Status checker should exit quietly.
+      }
+    }
+  }
+
+  @Override
+  public File tempDir()
+  {
+    return toolbox.getIndexingTmpDir();
+  }
+
+  @Override
+  public ControllerClient makeControllerClient(String controllerId)
+  {
+    final ServiceLocator locator = makeControllerLocator(controllerId);
+
+    return new IndexerControllerClient(
+        clientFactory.makeClient(
+            controllerId,
+            locator,
+            new SpecificTaskRetryPolicy(controllerId, StandardRetryPolicy.unlimited())
+        ),
+        jsonMapper(),
+        locator
+    );
+  }
+
+  @Override
+  public WorkerClient makeWorkerClient()
+  {
+    // Ignore workerId parameter. The workerId is passed into each method of WorkerClient individually.
+    return new IndexerWorkerClient(clientFactory, makeOverlordClient(), jsonMapper());
+  }
+
+  @Override
+  public FrameContext frameContext(QueryDefinition queryDef, int stageNumber)
+  {
+    final int numWorkersInJvm;
+
+    // Determine the max number of workers in JVM for memory allocations.
+    if (toolbox.getAppenderatorsManager() instanceof UnifiedIndexerAppenderatorsManager) {
+      // CliIndexer
+      numWorkersInJvm = injector.getInstance(WorkerConfig.class).getCapacity();
```

Comment:
i wonder if longer term this makes sense to be its own config? sort of seems to be giving a double meaning, though close enough in purpose that is fine

Reply:
This code is using the appropriate config: it's trying to figure out how many tasks could run in the same JVM. So a new config property isn't needed. However, a new injectable object (or a new method on an existing object) would remove the need for an `instanceof` here.